### PR TITLE
Use UMD bundle and ES6 modules

### PR DESCRIPTION
**Pull request by kamilkisiela**
UMD bundle for everyone and a version with ES6 modules for those who might be interested in tree-shaking etc.

@helfer We talked some time ago about the tree-shaking. I already moved `angular2-apollo` (released), `apollo-client-rxjs` (not yet published) and created an example (will publish it later today).

TODO:

- [x] If this PR is a new feature, reference an issue where a consensus about the design was reached (not necessary for small changes)
- [x] Make sure all of the significant new logic is covered by tests
- [x] Rebase your changes on master so that they can be merged easily
- [x] Make sure all tests and linter rules pass
- [x] Update CHANGELOG.md with your change
- [x] Add your name and email to the AUTHORS file (optional)
- [x] If this was a change that affects the external API, update the docs and post a link to the PR in the discussion

---

## Rollup

We use `rollup` to produce the UMD bundle that contains all files. This way we get smaller size of the library and support for CommonJS and AMD.

## ESM (ES6 Modules)

We compile TS files to ES5, so this stays the same. Only difference is that we use es6 modules instead of commonjs. Thanks to that change the `apollo-client` is tree-shaking friendly

## `main` and `module`

To make these changes to work with most of build tools we use `main` and `module` fields of `package.json`:

```json
{
  "main": "apollo.umd.js",
  "module": "index.js"
}
```

`module` tells where to look for ES6 modules.
`main` the same as before, points to files with CommonJS (in this case, only one file)

## Filesize

Because we use `rollup` and we have everything inside just one file, size of the library has been reduced.

Before:
The total size of index.min.js is **136 kB**.
The total gzipped size of index.min.js is **35.7 kB**.

After:
The total size of index.min.js is **127 kB**.
The total gzipped size of index.min.js is **34.8 kB**.

## Tests

You probably noticed that I use `babel` to compile already compiled files.

When TypeScript sees this:

```ts
import isBoolean from 'lodash/isBoolean';
console.log(isBoolean(false));
```

It's being compiled to:

```js
var isBoolean = require('lodash/isBoolean');
console.log(isBoolean(false).default);
```

And since `lodash/isBoolean` looks like this:

```js
var isBoolean = require('../index').isBoolean;
module.exports = isBoolean;
```

We end up with `isBoolean` as `undefined`.

This is why I use `babel` to compile ES5 file with ES6 modules to CommonJS so `mocha` could handle it.

## Deploy

I also applied these changes in `scripts/deploy.sh`.

## Circular Dependencies

We also fixed the issue #1074 

---

After this PR we can merge these:

- kamilkisiela/apollo-client-rxjs#16
- apollostack/angular2-apollo#206

PS @Urigo I wrote a summary so everyone can easily follow changes etc

### CODE REVIEW

**Comment by helfer** [comment]
@kamilkisiela is this PR intended for reference only?

**Comment by kamilkisiela** [comment]
@helfer I was working on an example for `angular2-apollo` that has the tree-shaking and the Ahead-of-Time compilation and it would be nice to have es6 modules in `apollo-client` to make the app even smaller and tree-shake friendly.

So I ended up with these changes and I decided to open a PR and a discussion :)

**Comment by helfer** [comment]
@kamilkisiela @Urigo this is a pretty major change, so it would have been very useful if you let me know beforehand so we could plan things and avoid wasting time by having to redo parts of this PR (or other PRs) after merging other stuff. It's also unfortunate that removing circular dependencies was rolled in with this PR, I think we could have merged it more quickly otherwise (no need to manually check that the Apollo Client package doesn't break).

Other than that, I think this is pretty cool, so thanks for the hard work! It looks like tree-shaking doesn't actually do much for our library size, but it's always nice to support it.

File: package.json [hunk 1]
```diff
@@ -2,19 +2,23 @@
...
   "description": "A simple yet functional GraphQL client.",
-  "main": "./lib/src/index.js",
+  "main": "./lib/apollo.umd.js",
+  "module": "./lib/src/index.js",
   "typings": "./lib/src/index.d.ts",
   "scripts": {
     "dev": "./scripts/dev.sh",
-    "deploy": "rm -rf ./lib && npm run compile && ./scripts/deploy.sh",
-    "pretest": "npm run compile",
+    "deploy": "rm -rf ./lib && ./scripts/deploy.sh",
```
Comment (helfer) [inline package.json]: Is the assumption here that the compiler and bundler already ran at this point?

Comment (kamilkisiela) [inline package.json]: `npm run compile` is [here](https://github.com/apollostack/apollo-client/blob/453e1bf53698e0274a7b9a81c9113992171ffdce/scripts/deploy.sh#L10), so there's no point of running it twice

**Comment by DxCx** [comment]
Nicely done!
@helfer, please note that the library size kept about the same as not what you should look at.
With the patch apollo client will be treeshake ready, so clients that needs to use only part of the library will get much smaller bundle sizes..

File: package.json (hunk 1 above)
Comment (helfer) [inline package.json]: Makes sense.

**Comment by helfer** [comment]
@DxCx yeah, I understand that. At the same time it's surprising to see that tree shaking doesn't reduce the library size because I assumed that we had a lot of dependencies of which we only use a small part. I guess that means we were pretty good at only importing what we really needed.

**Comment by helfer** [comment]
@kamilkisiela Can I bug you again when I think things are stable enough to merge this into the 0.6 branch? A lot of conflicts will need to be resolved, so I think it would make sense to do this in a relatively short timeframe once we're ready.

**Comment by kamilkisiela** [comment]
@helfer Sure thing! :+1:

**Comment by helfer** [comment]
@kamilkisiela do you have time to change the base of this PR to zero-decimal-six in the next day or two? If you can, then I can merge it and release it with 0.6 :)

**Comment by kamilkisiela** [comment]
@helfer Sure thing. There's an issue with code coverage reports, I'm working on it.

**Comment by kamilkisiela** [comment]
@helfer Maybe we could update the `redux` to at least `^3.4.0`? This way we get [the official typescript definitions](https://github.com/reactjs/redux/releases/tag/v3.4.0) so we can drop `@types/redux`

**Comment by kamilkisiela** [comment]
Closes #746

**Comment by kamilkisiela** [comment]
> change the base of this PR to zero-decimal-six

@helfer Okay, done.

**Comment by helfer** [comment]
@kamilkisiela re updating redux: sure thing, feel free to update it to the latest (compatible) version!

**Comment by helfer** [comment]
> Okay, done.
Hm, that's weird, it still says the base is master on github.

**Comment by kamilkisiela** [comment]
@helfer Yes... Look now

**Comment by helfer** [comment]
Thanks a lot @kamilkisiela, merging now!

**Comment by lucasconstantino** [comment]
I'm trying to upgrade from 0.5.10 to 0.8.0 (as a sugestion from https://github.com/apollostack/apollo-client/issues/1194) but I'm experiencing an unexpected token error. As far as I could investigate, importing apollo-client as `import Apollo from "apollo-client"` and using Webpack to build my app is ending up with a `export function` somewhere in the bundled file.

Any thoughts?

**Pull request merged**